### PR TITLE
Improve error handling

### LINKED ISSUE #102: Exceptions thrown in Converter logic should only print the error message
## ❓ Problem or use case

When an exception is thrown from a Converter class the complete usage help of the underlying command will be printed in addition to the error message. 

## 👍 Preferred solution or suggestions

Only the error message should be printed when a Converter class throws an exception.

## 📜 Example

Use connect with private key but entering a wrong password:
```
mqtt> con -i gitseti --key mqtt-client-key-2.pem
Enter private key password:
Invalid value for option '--key': cannot convert 'mqtt-client-key-2.pem' to PrivateKey (java.lang.IllegalArgumentException: The password for decrypting the private key was wrong.)

Usage:  mqtt shell con [OPTIONS]

Connects an mqtt client

Options:
      --help              display this help message
      -se, --sessionExpiryInterval <sessionExpiryInterval>
                          The lifetime of the session of the connected client'
      -up, --userProperty <connectUserProperties>
                          A user property of the connect message
  -d, --debug             Enable debug mode
  -v, --verbose           Enable verbose mode
  -V, --mqttVersion <version>
                          The mqtt version used by the client (default: 5)
  -h, --host <host>       The hostname of the message broker (default 'localhost')
  -p, --port <port>       The port of the message broker (default: 1883)
  -i, --identifier <identifier>
                          The client identifier UTF-8 String (default randomly generated string)
      -ip, --identifierPrefix <identifierPrefix>
                          The prefix of the client Identifier UTF-8 String
  -u, --user <user>       The username for authentication
      -pw, --password [<password>]
                          The password for authentication
  -k, --keepAlive <keepAlive>
                          A keep alive of the client (in seconds) (default: 60)
  -c, --[no-]cleanStart   Define a clean start for the connection (default: true)
  -s, --secure            Use default ssl configuration if no other ssl options are specified (default: false)
      --cafile FILE       Path to a file containing trusted CA certificates to enable encrypted certificate based communication
      --capath DIR        Path to a directory containing certificate files to import to enable encrypted certificate based communication
      --ciphers <cipherSuites>[:<cipherSuites>...]
                          The client supported cipher suites list in IANA format separated with ':'
      --tls-version <supportedTLSVersions>
                          The TLS protocol version to use (default: {'TLSv.1.2'})
      --cert <clientCertificate>
                          The client certificate to use for client side authentication
      --key <clientPrivateKey>
                          The path to the client private key for client side authentication
      -Wt, --willTopic <willTopic>
                          The topic of the will message
      -Wm, --willMessage <willMessage>
                          The payload of the will message
      -Wq, --willQualityOfService <willQos>
                          Quality of service level for the will message (default: 0)
      -Wr, --[no-]willRetain
                          Will message as retained message (default: false)
      -We, --willMessageExpiryInterval <willMessageExpiryInterval>
                          The lifetime of the will message in seconds (default: no message expiry)
      -Wd, --willDelayInterval <willDelayInterval>
                          The Server delays publishing the client's will message until the will delay has passed (default: 0)
      -Wpf, --willPayloadFormatIndicator <willPayloadFormatIndicator>
                          The payload format indicator of the will message
      -Wct, --willContentType <willContentType>
                          A description of the will message's content
      -Wrt, --willResponseTopic <willResponseTopic>
                          The topic name for the response message
      -Wcd, --willCorrelationData <willCorrelationData>
                          The correlation data of the will message
      -Wup, --willUserProperty <willUserProperties>
                          A user property of the will message
      --rcvMax <receiveMaximum>
                          The maximum amount of not acknowledged publishes with QoS 1 or 2 the client accepts from the server concurrently. (default: 65535)
      --sendMax <sendMaximum>
                          The maximum amount of not acknowledged publishes with QoS 1 or 2 the client send to the server concurrently. (default: 65535)
      --maxPacketSize <maximumPacketSize>
                          The maximum packet size the client accepts from the server. (default: 268435460)
      --sendMaxPacketSize <sendMaximumPacketSize>
                          The maximum packet size the client sends to the server. (default: 268435460)
      --topicAliasMax <topicAliasMaximum>
                          The maximum amount of topic aliases the client accepts from the server. (default: 0)
      --sendTopicAliasMax <sendTopicAliasMaximum>
                          The maximum amount of topic aliases the client sends to the server. (default: 16)
      --[no-]reqProblemInfo
                          The client requests problem information from the server. (default: true)
      --[no-]reqResponseInfo
                          The client requests response information from the server. (default: false)
mqtt> 
```

Better behavior:

```
mqtt> con -i gitseti --key mqtt-client-key-2.pem
Enter private key password:
The password for decrypting the private key was wrong.
mqtt>
```

### PR DESCRIPTION
**Pull request by gitseti**
**Motivation**

resolves #102 

When an error occurs in the cli the handling of printing that error which the user is confronted with may be confusing in the current state. Therefore, this PR should improve the error printing and should give the user a clear indication of what went wrong with the processed command. Also, this PR does some internal refactoring by adding an abstraction layer to differentiate between the command line used in the shell and the command line for executing commands without a shell context.

**Changes** 
* Refactor `CommandLine` instances into the `CommandLineProvider` class which handles getting command lines and holds the attributes for the command lines
* add abstract class `CommonErrorMessageHandler` which handles printing exceptions thrown inside the cli without printing the usage help which would be default picocli behavior
* Add `CommandErrorMessageHandler` which handles printing exceptions to the user which were thrown when using a command outside the shell mode
* Add `ShellErrorMessageHandler` which handles printing exceptions to the user which were thrown inside the shell mode
* Add badges to README
[GitHub Release](https://github.com/hivemq/mqtt-cli/releases) 
[picocli](https://github.com/hivemq/mqtt-cli/releases)
[picocli](https://github.com/hivemq/mqtt-cli/blob/develop/LICENSE)
[picocli](https://github.com/hivemq/hivemq-mqtt-client)
[picocli](https://github.com/remkop/picocli)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,8 +1,17 @@
-# MQTT Command Line Interface
+<p align="center">
+  <a href="https://terminalizer.com">
```
Comment (SgtSilvio) [inline README.md]: Why this link, copy and paste error?

File: src/main/java/com/hivemq/cli/commands/shell/ShellCommand.java [hunk 2]
```diff
@@ -282,15 +279,14 @@ static boolean isDebug() {
         return DEBUG;
     }
 
-
-    @Override
-    public String toString() {
+@Override
+public String toString() {
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/cli/commands/shell/ShellCommand.java]: strange formatting

File: src/main/java/com/hivemq/cli/commands/shell/ShellCommand.java [hunk 3]
```diff
@@ -243,7 +240,7 @@ static void readFromShell() {
 
 
     static void usage(Object command) {
-        currentCommandLine.usage(command, System.out, MqttCLIMain.COLOR_SCHEME);
+        currentCommandLine.usage(command, System.out, CommandLineProvider.getColorScheme());
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/cli/commands/shell/ShellCommand.java]: Is this necessary, isn't there a printUsage() method or something similar?

File: src/main/java/com/hivemq/cli/commandline/CommandLineProvider.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+ *
+ */
+package com.hivemq.cli.commandline;
+
+import com.hivemq.cli.ioc.DaggerContextCommandLine;
+import com.hivemq.cli.ioc.DaggerMqttCLI;
+import org.jetbrains.annotations.NotNull;
+import picocli.CommandLine;
+
+public class CommandLineProvider {
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/cli/commandline/CommandLineProvider.java]: Imho this provider should be handled with Dagger and the color scheme, etc. should be configured in the modules provide methods.

File: src/main/java/com/hivemq/cli/commands/shell/ShellCommand.java (hunk 3 above)
Comment (gitseti) [reply src/main/java/com/hivemq/cli/commands/shell/ShellCommand.java]: This is a wrapper for always printing with the currently active command line in the shell mode. 

There is actually no such method but an alternative would be:
`System.out.println(currentCommandLine.getUsageMessage(CommandLineProvider.getColorScheme()));`

What do you think about this solution?

Comment (SgtSilvio) [reply src/main/java/com/hivemq/cli/commands/shell/ShellCommand.java]: Then it is ok, I just wondered if why you have to provide the color scheme, although it is already set for the currentCommandLine.

Comment (gitseti) [reply src/main/java/com/hivemq/cli/commands/shell/ShellCommand.java]: Actually I think you're right and this gets solved with your proposed Dagger changes. 👍

File: src/main/java/com/hivemq/cli/commandline/CommandLineConfig.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ */
+package com.hivemq.cli.commandline;
+
+import org.jetbrains.annotations.NotNull;
+import picocli.CommandLine;
+
+import javax.inject.Inject;
+
+public class CommandLineConfig {
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/cli/commandline/CommandLineConfig.java]: Singleton or use static methods

File: src/main/java/com/hivemq/cli/commandline/CommonErrorMessageHandler.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+import org.jetbrains.annotations.NotNull;
+import picocli.CommandLine;
+
+import java.io.PrintWriter;
+
+abstract class CommonErrorMessageHandler implements CommandLine.IParameterExceptionHandler {
+
+    CommandLine cmd;
+    PrintWriter writer;
+    private CommandLine.ParameterException parameterException;
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/cli/commandline/CommonErrorMessageHandler.java]: Fields should not be used for storing context information, please provide them as arguments to the methods.

File: src/main/java/com/hivemq/cli/commands/shell/ShellCommand.java [hunk 7]
```diff
@@ -127,12 +128,8 @@ public void run() {
 
 
     private void interact() {
-        shellCommandLine = new CommandLine(spec);
-        contextCommandLine = DaggerContextCommandLine.create().contextCommandLine();
-
-        shellCommandLine.setColorScheme(MqttCLIMain.COLOR_SCHEME);
-        contextCommandLine.setColorScheme(MqttCLIMain.COLOR_SCHEME);
-        contextCommandLine.setUsageHelpWidth(MqttCLIMain.CLI_WIDTH);
+        shellCommandLine = DaggerMqttCLI.create().commandLine().getSubcommands().get("shell");
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/cli/commands/shell/ShellCommand.java]: I don't think it is intended to create a new Dagger component and a new command line here, is it?

Comment (SgtSilvio) [reply src/main/java/com/hivemq/cli/commands/shell/ShellCommand.java]: maybe use a static field in MqttCLI which stores `DaggerMqttCLI.create()`

File: src/main/java/com/hivemq/cli/commandline/ShellErrorMessageHandler.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+public class ShellErrorMessageHandler extends CommonErrorMessageHandler implements CommandLine.IParameterExceptionHandler {
+
+    @Inject ShellErrorMessageHandler() {};
+
+    @Override
+    public int handleParseException(final @NotNull CommandLine.ParameterException ex, final @NotNull String[] args) throws Exception {
+       int exitCode = super.handleParseException(ex, args);
+
+        final PrintWriter writer = ex.getCommandLine().getErr();
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/cli/commandline/ShellErrorMessageHandler.java]: formatting

File: src/main/java/com/hivemq/cli/commandline/ShellErrorMessageHandler.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+       if (ex instanceof CommandLine.UnmatchedArgumentException &&
+               ((CommandLine.UnmatchedArgumentException) ex).getUnmatched().get(0).equals(args[0])) {
+           writer.printf("Try 'help' to get a list of commands.%n");
+       }
+       else {
+           writer.printf("Try 'help %s' for more information.%n", args[0]);
+        }
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/cli/commandline/ShellErrorMessageHandler.java]: formatting

**Pull request merged**